File: lib/dapa_bids_proxy.py
"""방위사업청 군수품조달 입찰공고 — data.go.kr BidPblancInfoService"""

from __future__ import annotations

import os
import re
import urllib.error
import urllib.parse
import urllib.request
import xml.etree.ElementTree as ET
from datetime import date, datetime, timedelta

BASE_URL = "https://apis.data.go.kr/1690000/BidPblancInfoService/getDmstcCmpetBidPblancList"
D2B_DETAIL_URL = "https://www.d2b.go.kr/pdb/bid/bidDetail.do"

_bids_cache: tuple[float, dict] | None = None
BIDS_CACHE_SEC = 300


def get_data_go_kr_key() -> str:
    return os.environ.get("DATA_GO_KR_SERVICE_KEY", "").strip().strip('"').strip("'")


def _fmt_yyyymmdd(d: date) -> str:
    return d.strftime("%Y%m%d")


def _local_tag(tag: str) -> str:
    return tag.split("}")[-1] if "}" in tag else tag


def _text(el: ET.Element | None) -> str:
    if el is None or el.text is None:
        return ""
    return el.text.strip()


def _parse_items(xml_bytes: bytes) -> list[dict]:
    root = ET.fromstring(xml_bytes)
    result_code = ""
    result_msg = ""

    for el in root.iter():
        tag = _local_tag(el.tag)
        if tag == "resultCode":
            result_code = _text(el)
        elif tag == "resultMsg":
            result_msg = _text(el)

    if result_code and result_code not in {"00", "0"}:
        raise RuntimeError(result_msg or f"API error ({result_code})")

    items: list[dict] = []
    for el in root.iter():
        if _local_tag(el.tag) != "item":
            continue
        row = {_local_tag(child.tag): _text(child) for child in el}
        if row:
            items.append(row)

    return items


def _parse_dt(raw: str) -> datetime | None:
    if not raw:
        return None
    s = re.sub(r"\D", "", raw)
    try:
        if len(s) >= 14:
            return datetime.strptime(s[:14], "%Y%m%d%H%M%S")
        if len(s) >= 8:
            return datetime.strptime(s[:8], "%Y%m%d")
    except ValueError:
        pass
    for fmt in ("%Y-%m-%d %H:%M:%S", "%Y-%m-%d"):
        try:
            return datetime.strptime(raw.strip(), fmt)
        except ValueError:
            continue
    return None


def _deadline_label(dt: datetime | None) -> tuple[str, str]:
    if not dt:
        return "—", "open"
    days = int((dt - datetime.now()).total_seconds() // 86400)
    if days < 0:
        return "마감", "closed"
    if days <= 7:
        return f"D-{days}", "urgent"
    return f"D-{days}", "open"


def _bid_url(item: dict) -> str:
    bidno = item.get("g2bPblancNo") or item.get("pblancNo") or ""
    bidseq = item.get("g2bPblancOdr") or item.get("pblancOdr") or "00"
    if bidno:
        qs = urllib.parse.urlencode({"bidno": bidno, "bidseq": bidseq})
        return f"{D2B_DETAIL_URL}?{qs}"
    return "https://www.d2b.go.kr/"


def _map_bid(item: dict, index: int) -> dict:
    close_raw = (
        item.get("biddocPresentnClosDt")
        or item.get("bidPartcptRegistClosDt")
        or item.get("opengDt")
        or ""
    )
    close_dt = _parse_dt(close_raw)
    deadline, status = _deadline_label(close_dt)

    pblanc_no = item.get("pblancNo") or item.get("g2bPblancNo") or f"DAPA-{index + 1}"
    meta_parts = [p for p in [item.get("busiDivs"), item.get("cntrctMth")] if p]
    budget = " · ".join(meta_parts) if meta_parts else "공고문 확인"

    return {
        "id": pblanc_no,
        "title": item.get("bidNm") or "제목 없음",
        "agency": item.get("ornt") or "방위사업청",
        "budget": budget,
        "deadline": deadline,
        "status": status,
        "url": _bid_url(item),
        "pblancDate": item.get("pblancDate") or "",
        "opengDt": item.get("opengDt") or "",
        "closeDt": close_raw,
    }


def _sort_bids(bids: list[dict]) -> list[dict]:
    def sort_key(b: dict) -> tuple:
        close = _parse_dt(b.get("closeDt") or "") or datetime.max
        pub = _parse_dt(b.get("pblancDate") or "") or datetime.min
        urgent = 0 if b.get("status") == "urgent" else 1
        return (urgent, close, -pub.timestamp())

    return sorted(bids, key=sort_key)


def fetch_dapa_bids(
    *,
    page_no: int = 1,
    num_of_rows: int = 10,
    days_back: int = 30,
    http_get_bytes,
) -> dict:
    global _bids_cache

    import time

    now = time.time()
    if _bids_cache and now - _bids_cache[0] < BIDS_CACHE_SEC:
        return _bids_cache[1]

    api_key = get_data_go_kr_key()
    if not api_key:
        raise RuntimeError("DATA_GO_KR_SERVICE_KEY is not configured on the server")

    end = date.today()
    start = end - timedelta(days=max(1, days_back) - 1)

    params = {
        "pageNo": str(page_no),
        "numOfRows": str(min(max(num_of_rows, 1), 100)),
        "anmtDateBegin": _fmt_yyyymmdd(start),
        "anmtDateEnd": _fmt_yyyymmdd(end),
    }
    qs = urllib.parse.urlencode(params)
    url = f"{BASE_URL}?{qs}&serviceKey={api_key}"

    try:
        xml_bytes = http_get_bytes(url, timeout=20)
    except urllib.error.HTTPError as e:
        raw = e.read().decode("utf-8", errors="replace")
        raise RuntimeError(raw or f"HTTP {e.code}") from e
    except urllib.error.URLError as e:
        raise RuntimeError(f"Connection failed: {e.reason}") from e

    items = _parse_items(xml_bytes)
    bids = _sort_bids([_map_bid(item, i) for i, item in enumerate(items)])

    result = {
        "bids": bids,
        "totalCount": len(bids),
        "source": "dapa",
        "live": True,
        "query": {
            "anmtDateBegin": params["anmtDateBegin"],
            "anmtDateEnd": params["anmtDateEnd"],
        },
    }
    _bids_cache = (now, result)
    return result

File: lib/dapa-bids-proxy.js
/**
 * 방위사업청 군수품조달 입찰공고 — data.go.kr BidPblancInfoService
 */

const BASE_URL =
  "https://apis.data.go.kr/1690000/BidPblancInfoService/getDmstcCmpetBidPblancList";
const D2B_DETAIL_URL = "https://www.d2b.go.kr/pdb/bid/bidDetail.do";

let bidsCache = null;
let bidsCacheAt = 0;
const BIDS_CACHE_MS = 5 * 60 * 1000;

function getDataGoKrKey() {
  const raw = process.env.DATA_GO_KR_SERVICE_KEY || "";
  return raw.trim().replace(/^["']|["']$/g, "");
}

function fmtYyyymmdd(d) {
  const y = d.getFullYear();
  const m = String(d.getMonth() + 1).padStart(2, "0");
  const day = String(d.getDate()).padStart(2, "0");
  return `${y}${m}${day}`;
}

function localTag(tag) {
  const i = tag.indexOf("}");
  return i >= 0 ? tag.slice(i + 1) : tag;
}

function parseItems(xml) {
  const codeMatch = xml.match(/<resultCode>([^<]*)<\/resultCode>/);
  const msgMatch = xml.match(/<resultMsg>([^<]*)<\/resultMsg>/);
  const resultCode = codeMatch?.[1]?.trim() || "";
  const resultMsg = msgMatch?.[1]?.trim() || "";

  if (resultCode && resultCode !== "00" && resultCode !== "0") {
    const err = new Error(resultMsg || `API error (${resultCode})`);
    err.status = 502;
    throw err;
  }

  const items = [];
  const blocks = xml.match(/<item>[\s\S]*?<\/item>/g) || [];
  for (const block of blocks) {
    const row = {};
    const fields = block.matchAll(/<([a-zA-Z0-9_]+)>([^<]*)<\/\1>/g);
    for (const [, name, value] of fields) {
      if (name !== "item") row[name] = value.trim();
    }
    if (Object.keys(row).length) items.push(row);
  }
  return items;
}

function parseDt(raw) {
  if (!raw) return null;
  const digits = String(raw).replace(/\D/g, "");
  if (digits.length >= 14) {
    const d = new Date(
      Number(digits.slice(0, 4)),
      Number(digits.slice(4, 6)) - 1,
      Number(digits.slice(6, 8)),
      Number(digits.slice(8, 10)),
      Number(digits.slice(10, 12)),
      Number(digits.slice(12, 14))
    );
    return Number.isNaN(d.getTime()) ? null : d;
  }
  if (digits.length >= 8) {
    const d = new Date(
      Number(digits.slice(0, 4)),
      Number(digits.slice(4, 6)) - 1,
      Number(digits.slice(6, 8))
    );
    return Number.isNaN(d.getTime()) ? null : d;
  }
  const d = new Date(raw);
  return Number.isNaN(d.getTime()) ? null : d;
}

function deadlineLabel(dt) {
  if (!dt) return { deadline: "—", status: "open" };
  const days = Math.ceil((dt.getTime() - Date.now()) / 86400000);
  if (days < 0) return { deadline: "마감", status: "closed" };
  if (days <= 7) return { deadline: `D-${days}`, status: "urgent" };
  return { deadline: `D-${days}`, status: "open" };
}

function bidUrl(item) {
  const bidno = item.g2bPblancNo || item.pblancNo || "";
  const bidseq = item.g2bPblancOdr || item.pblancOdr || "00";
  if (bidno) {
    return `${D2B_DETAIL_URL}?bidno=${encodeURIComponent(bidno)}&bidseq=${encodeURIComponent(bidseq)}`;
  }
  return "https://www.d2b.go.kr/";
}

function mapBid(item, index) {
  const closeRaw =
    item.biddocPresentnClosDt || item.bidPartcptRegistClosDt || item.opengDt || "";
  const closeDt = parseDt(closeRaw);
  const { deadline, status } = deadlineLabel(closeDt);
  const meta = [item.busiDivs, item.cntrctMth].filter(Boolean);

  return {
    id: item.pblancNo || item.g2bPblancNo || `DAPA-${index + 1}`,
    title: item.bidNm || "제목 없음",
    agency: item.ornt || "방위사업청",
    budget: meta.length ? meta.join(" · ") : "공고문 확인",
    deadline,
    status,
    url: bidUrl(item),
    pblancDate: item.pblancDate || "",
    opengDt: item.opengDt || "",
    closeDt: closeRaw,
  };
}

function sortBids(bids) {
  return [...bids].sort((a, b) => {
    const au = a.status === "urgent" ? 0 : 1;
    const bu = b.status === "urgent" ? 0 : 1;
    if (au !== bu) return au - bu;
    const ac = parseDt(a.closeDt)?.getTime() ?? Number.MAX_SAFE_INTEGER;
    const bc = parseDt(b.closeDt)?.getTime() ?? Number.MAX_SAFE_INTEGER;
    if (ac !== bc) return ac - bc;
    const ap = parseDt(a.pblancDate)?.getTime() ?? 0;
    const bp = parseDt(b.pblancDate)?.getTime() ?? 0;
    return bp - ap;
  });
}

async function fetchDapaBids(options = {}) {
  if (bidsCache && Date.now() - bidsCacheAt < BIDS_CACHE_MS) {
    return bidsCache;
  }

  const apiKey = getDataGoKrKey();
  if (!apiKey) {
    const err = new Error("DATA_GO_KR_SERVICE_KEY is not configured on the server");
    err.status = 500;
    throw err;
  }

  const pageNo = options.pageNo || 1;
  const numOfRows = Math.min(Math.max(options.numOfRows || 10, 1), 100);
  const daysBack = options.daysBack || 30;

  const end = new Date();
  const start = new Date();
  start.setDate(end.getDate() - (Math.max(daysBack, 1) - 1));

  const url = new URL(BASE_URL);
  url.searchParams.set("pageNo", String(pageNo));
  url.searchParams.set("numOfRows", String(numOfRows));
  url.searchParams.set("anmtDateBegin", fmtYyyymmdd(start));
  url.searchParams.set("anmtDateEnd", fmtYyyymmdd(end));
  const requestUrl = `${url.toString()}&serviceKey=${encodeURIComponent(apiKey)}`;

  const res = await fetch(requestUrl, {
    headers: { "User-Agent": "LIG-Dashboard/1.0 (dapa-bids-proxy)" },
  });
  const xml = await res.text();
  if (!res.ok) {
    const err = new Error(xml || `HTTP ${res.status}`);
    err.status = res.status;
    throw err;
  }

  const items = parseItems(xml);
  const bids = sortBids(items.map((item, i) => mapBid(item, i)));

  bidsCache = {
    bids,
    totalCount: bids.length,
    source: "dapa",
    live: true,
    query: {
      anmtDateBegin: fmtYyyymmdd(start),
      anmtDateEnd: fmtYyyymmdd(end),
    },
  };
  bidsCacheAt = Date.now();
  return bidsCache;
}

module.exports = { fetchDapaBids, getDataGoKrKey };
